workflow-trace-viewer/src/jvmMain/kotlin/com/squareup/workflow1/traceviewer/ui/WorkflowTree.kt

```diff
@@ -29,16 +30,18 @@ import io.github.vinceglb.filekit.PlatformFile
  * tabs. This will also all errors related to errors parsing a given trace JSON file.
  */
 @Composable
-public fun RenderDiagram(
+internal fun RenderDiagram(
   traceFile: PlatformFile,
   frameInd: Int,
   onFileParse: (List<Node>) -> Unit,
-  onNodeSelect: (Node) -> Unit,
+  onNodeSelect: (Node, Node?) -> Unit,
   modifier: Modifier = Modifier
 ) {
-  var frames by remember { mutableStateOf<List<Node>>(emptyList()) }
   var isLoading by remember(traceFile) { mutableStateOf(true) }
   var error by remember(traceFile) { mutableStateOf<Throwable?>(null) }
+  var frames by remember { mutableStateOf<List<Node>>(emptyList()) }
+  var fullTree by remember { mutableStateOf<List<Node>>(emptyList()) }
+  var affectedNodes by remember { mutableStateOf<List<Set<Node>>>(emptyList()) }
```

Comment:
Nit: Can any of these be `SnapshotStateList`s? 

Or if you want them to be immutable, then consider using `ImmutableList` and friends.